### PR TITLE
support union and interface types in remote schema (close #1276)

### LINKED ISSUE #1276: support interface and union types for remote graphql schemas
Currently if remote schemas has interfaces, hasura throws an error.

### PR DESCRIPTION
**Pull request by nizar-m**
### Description

What component does this PR affect? 

- [x] Server
- [ ] Console
- [ ] CLI
- [ ] Docs
- [ ] Community Content
- [ ] Build System

Requires changes from other components? If yes, please mark the components:

- [ ] Server
- [ ] Console
- [ ] CLI
- [ ] Docs
- [ ] Community Content
- [ ] Build System

### Related Issue
#1276 

### Solution and Design

- Validate implementation of interfaces by object types
- Validate union types
- Support introspection query for union and interface

### Type

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Docs update
- [ ] Community content

### Checklist:

- [ ] I have read the **[contributing guide](https://github.com/hasura/graphql-engine/blob/master/CONTRIBUTING.md)** and my code conforms to the guidelines.
- [ ] This change requires a change in the documentation. 
- [ ] I have updated the documentation accordingly.
- [x] I have added required tests.

### CODE REVIEW

File: server/src-lib/Hasura/GraphQL/Validate/Types.hs [hunk 1]
```diff
@@ -156,17 +161,23 @@ instance EquatableGType ObjTyInfo where
   getEqProps a = (,) (_otiName a) (Map.map getEqProps (_otiFields a))
```
Comment (0x777) [inline server/src-lib/Hasura/GraphQL/Validate/Types.hs]: We'll also need to check if the interfaces that are implemented are same.

File: server/src-lib/Hasura/GraphQL/Validate/Types.hs [hunk 2]
```diff
@@ -145,9 +149,10 @@ type ObjFieldMap = Map.HashMap G.Name ObjFldInfo
 
 data ObjTyInfo
   = ObjTyInfo
-  { _otiDesc   :: !(Maybe G.Description)
-  , _otiName   :: !G.NamedType
-  , _otiFields :: !ObjFieldMap
+  { _otiDesc       :: !(Maybe G.Description)
+  , _otiName       :: !G.NamedType
+  , _otiImplIFaces :: [G.NamedType]
```
Comment (0x777) [inline server/src-lib/Hasura/GraphQL/Validate/Types.hs]: Maybe we can use a hashset here?

File: server/src-lib/Hasura/GraphQL/Validate/Types.hs (hunk 1 above)
Comment (nizar-m) [reply server/src-lib/Hasura/GraphQL/Validate/Types.hs]: Changed getEqProps to output a 3 part tuple, which includes HashSet of implementing interfaces

File: server/src-lib/Hasura/GraphQL/Validate/Types.hs (hunk 2 above)
Comment (nizar-m) [reply server/src-lib/Hasura/GraphQL/Validate/Types.hs]: Sure. Changed from Array of implementing interfaces to HashSet.

**Pull request merged**